### DISCUSSION TITLE
Best Practice for forms using Riverpod

### DISCUSSION THREAD

**Discussion by TreyThomas93** [rrousselGit/riverpod#3476] [Q&A]
I recently created a small project in Github to demonstrate some approaches that I take in using forms with Riverpod.
[Repo](https://github.com/TreyThomas93/flutter_form_examples)

There are 3 examples that I made. Example 1 is a basic approach using texteditingcontrollers. nothing fancy. but to illustrate form state management at its most basic.
Examples 2 and 3 use a Notifier provider to hold the state of a form model of some type. Then I use this throughout the form. I still use the form for validation. But then I am at difference on how I should handle the submission of the form. If I handle it through a method within the form state notifier, then I have to do exception handling within the calling widget. Doing this doesnt allow me to funnel any exceptions to the provider observer I am using, for which I prefer it that way for all my logging. Therefore, example 3 is my go to typically. It holds the form state in a notifier, but the submission goes through another notifier that uses a sealed union class as its state, for which a ref.listen updates the ui (loading circle, snackbar, navigate to another screen, ect) based on the submission state. And then I can capture the error in the provider observer if of type "ErrorBase" or whatever and log it from there. I typically go this route for synchronous submissions, but I still go this route for async subs as well. Although I try to use AsyncValue wherever its the most logical for me. 

Any advice on best practices or better ways to go about this is welcomed. Thank you.

**Comment by philitell** [comment]
Great Job - i like the examples 3 and 4.
I have used a similar approach - the only difference is that i combined "FormState" and "FormController" in one single ChangeNotifier (which is not recommended anymore). 
Thanks for sharing your examples!
